juju/juju
# Tick the Raft lease clock directly instead of via pub/sub

Juju runs a clock updater worker that ticks the internal clock of the Raft FSM.

Despite ensuring via manifold config that this worker runs on the Raft leader node, we still use the lease store implementation of clock advancement, which issues commands over pub/sub and in turn via the raft forwarder. This is superfluous round-tripping.

Here, we update the worker to eschew dependency on the lease manager, and instead receive a `raft.Raft` instance and the `FSM`, so that we can apply clock advancement operations directly to Raft.

## QA steps

- Bootstrap and `enable-ha`.
- Deploy some multi-unit workloads.
- Stop one of the unit agents somehow to cause leadership to change, and observe the absence of logged errors.
- Cause the Raft leader to change and repeat the step above to ensure that we keep ticking on the leader.

## Documentation changes

None.

## Bug reference

N/A
